File: packages/typescript/src/hooks/telemetry-hook.ts
import type { Hook, HookContext, EvaluationDetails, FlagValue, HookHints, BeforeHookContext, ErrorCode } from '@openfeature/server-sdk';

/**
 * Telemetry event data
 */
export interface TelemetryEvent {
	type: 'evaluation' | 'error';
	flagKey: string;
	timestamp: number;
	duration?: number;
	value?: unknown;
	reason?: string;
	variant?: string;
	/** The OpenFeature ErrorCode, if evaluation produced an error. */
	errorCode?: ErrorCode;
	errorMessage?: string;
	context?: Record<string, unknown>;
	/**
	 * Caller-supplied hints forwarded from `EvaluationOptions.hookHints`.
	 * Use this to attach trace IDs, request IDs, or any other metadata
	 * that the application author wants to associate with the evaluation.
	 */
	hints?: Readonly<Record<string, unknown>>;
}

/**
 * Telemetry hook for tracking flag evaluations
 *
 * @example
 * ```typescript
 * import { OpenFeature } from '@openfeature/server-sdk';
 * import { FlagshipServerProvider, TelemetryHook } from '@cloudflare/flagship/server';
 *
 * const telemetryHook = new TelemetryHook((event) => {
 *   // Send to your analytics service
 *   analytics.track('flag_evaluated', event);
 * });
 *
 * OpenFeature.addHooks(telemetryHook);
 * ```
 */
export class TelemetryHook implements Hook {
	// Maps each evaluation's unique key to its start timestamp.
	// Keyed by a string stored in `contextKeys` to link the hookContext to its entry.
	private startTimes: Map<string, number> = new Map();

	// WeakMap<object> avoids the `as HookContext` cast: OpenFeature passes the same
	// object reference to all hook stages, so the lookup is always correct at runtime.
	private contextKeys: WeakMap<object, string> = new WeakMap();
	private hints: WeakMap<object, HookHints> = new WeakMap();

	private readonly onEvent: (event: TelemetryEvent) => void;

	constructor(onEvent: (event: TelemetryEvent) => void) {
		this.onEvent = onEvent;
	}

	before(hookContext: BeforeHookContext, hookHints?: HookHints): void {
		const now = Date.now();
		const key = `${hookContext.flagKey}-${now}-${Math.random()}`;
		this.startTimes.set(key, now);
		this.contextKeys.set(hookContext, key);
		// Store hints alongside the key so they can be forwarded in after/error.
		if (hookHints !== undefined) {
			this.hints.set(hookContext, hookHints);
		}
	}

	after(hookContext: Readonly<HookContext>, evaluationDetails: EvaluationDetails<FlagValue>, _hookHints?: HookHints): void {
		const telemetryKey = this.contextKeys.get(hookContext);
		const startTime = telemetryKey ? this.startTimes.get(telemetryKey) : undefined;
		const duration = startTime !== undefined ? Date.now() - startTime : undefined;

		if (telemetryKey !== undefined) {
			this.startTimes.delete(telemetryKey);
			this.contextKeys.delete(hookContext);
		}

		this.onEvent({
			type: 'evaluation',
			flagKey: hookContext.flagKey,
			timestamp: Date.now(),
			duration,
			value: evaluationDetails.value,
			reason: evaluationDetails.reason,
			variant: evaluationDetails.variant,
			errorCode: evaluationDetails.errorCode,
			context: hookContext.context,
			hints: this.hints.get(hookContext),
		});
	}

	error(hookContext: Readonly<HookContext>, error: unknown, _hookHints?: HookHints): void {
		const telemetryKey = this.contextKeys.get(hookContext);
		const startTime = telemetryKey ? this.startTimes.get(telemetryKey) : undefined;
		const duration = startTime !== undefined ? Date.now() - startTime : undefined;

		if (telemetryKey !== undefined) {
			this.startTimes.delete(telemetryKey);
			this.contextKeys.delete(hookContext);
		}

		const errorMessage = error instanceof Error ? error.message : String(error);
		this.onEvent({
			type: 'error',
			flagKey: hookContext.flagKey,
			timestamp: Date.now(),
			duration,
			errorMessage,
			context: hookContext.context,
			hints: this.hints.get(hookContext),
		});
	}

	finally(hookContext: Readonly<HookContext>, _evaluationDetails: EvaluationDetails<FlagValue>, _hookHints?: HookHints): void {
		const telemetryKey = this.contextKeys.get(hookContext);
		if (telemetryKey !== undefined) {
			this.startTimes.delete(telemetryKey);
			this.contextKeys.delete(hookContext);
		}
		this.hints.delete(hookContext);
	}
}

File: sdks/python/src/flagship/hooks/telemetry_hook.py
import time
import weakref
from collections.abc import Callable, Mapping
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any, Literal

from openfeature.hook import Hook

if TYPE_CHECKING:
    from openfeature.exception import ErrorCode
    from openfeature.flag_evaluation import FlagEvaluationDetails, FlagValueType
    from openfeature.hook import HookContext, HookHints

__all__ = ["TelemetryEvent", "TelemetryHook"]


@dataclass
class TelemetryEvent:
    type: Literal["evaluation", "error"]
    flag_key: str
    timestamp: int
    duration_ms: float | None = None
    value: Any = None
    reason: str | None = None
    variant: str | None = None
    error_code: "ErrorCode | None" = None
    error_message: str | None = None
    context: Any = None
    hints: Mapping[str, Any] | None = None


class TelemetryHook(Hook):
    """Emits a :class:`TelemetryEvent` for every flag evaluation.

    Useful for sending evaluation metrics to an analytics service::

        api.add_hooks([TelemetryHook(analytics.track)])
    """

    def __init__(self, on_event: Callable[[TelemetryEvent], None]) -> None:
        self._on_event = on_event
        self._start_times: weakref.WeakKeyDictionary[Any, float] = weakref.WeakKeyDictionary()
        self._hints: weakref.WeakKeyDictionary[Any, Any] = weakref.WeakKeyDictionary()

    def before(self, hook_context: "HookContext", hints: "HookHints") -> None:
        self._start_times[hook_context] = time.monotonic()
        if hints:
            self._hints[hook_context] = hints

    def after(
        self,
        hook_context: "HookContext",
        details: "FlagEvaluationDetails[FlagValueType]",
        hints: "HookHints",
    ) -> None:
        self._on_event(
            TelemetryEvent(
                type="evaluation",
                flag_key=hook_context.flag_key,
                timestamp=_now_ms(),
                duration_ms=self._pop_duration(hook_context),
                value=details.value,
                reason=str(details.reason) if details.reason is not None else None,
                variant=details.variant,
                error_code=details.error_code,
                error_message=details.error_message,
                context=hook_context.evaluation_context,
                hints=self._hints.get(hook_context),
            )
        )

    def error(
        self,
        hook_context: "HookContext",
        exception: Exception,
        hints: "HookHints",
    ) -> None:
        self._on_event(
            TelemetryEvent(
                type="error",
                flag_key=hook_context.flag_key,
                timestamp=_now_ms(),
                duration_ms=self._pop_duration(hook_context),
                error_message=str(exception),
                context=hook_context.evaluation_context,
                hints=self._hints.get(hook_context),
            )
        )

    def finally_after(
        self,
        hook_context: "HookContext",
        details: "FlagEvaluationDetails[FlagValueType]",
        hints: "HookHints",
    ) -> None:
        self._start_times.pop(hook_context, None)
        self._hints.pop(hook_context, None)

    def _pop_duration(self, hook_context: "HookContext") -> float | None:
        start = self._start_times.pop(hook_context, None)
        return None if start is None else (time.monotonic() - start) * 1000.0


def _now_ms() -> int:
    return int(time.time() * 1000)
